src/test/java/com/google/sps/servlets/UserFoldersServletTest.java

```diff
@@ -0,0 +1,186 @@
+
+    List<Folder> noFoldersInDatastore = new ArrayList<>();
+    
+    // Generate testing User
+    Entity user = new Entity("User", "testId");
+    String userKey = KeyFactory.keyToString(user.getKey());
+
+    when(mockRequest.getParameter("userKey")).thenReturn(userKey);
+    servlet.doGet(mockRequest, mockResponse);
+    String response = responseWriter.toString();
+    String expectedResponse = new Gson().toJson(getExpectedJsonInfo(/*folder=*/noFoldersInDatastore, /*showCreateFormStatus=*/true));
+
+    assertTrue(compareJson(response, expectedResponse));
+  }
+
+  @Test
+  public void UserNotLoggedIn() throws Exception {
+    /* Returns an empty array list and signals front-end to not show Create Folder Option */
+
+    helper.setEnvIsLoggedIn(false);
+    List<Folder> noFoldersQueried = new ArrayList<>();
+
+    servlet.doGet(mockRequest, mockResponse);
+    String response = responseWriter.toString();
+    String expectedResponse = new Gson().toJson(getExpectedJsonInfo(/*folder=*/noFoldersQueried, /*showCreateFormStatus=*/false));
+
+    assertTrue(compareJson(response, expectedResponse));
+  }
+
+  @Test
+  public void UserCreatesFirstFolder() throws Exception {
+    /* First time user is creating a folder, so current number of folders should be 1 */
+    
+    // Generate testing User
+    Entity user = new Entity("User", "testId");
+    String userKey = KeyFactory.keyToString(user.getKey());
+
+    when(mockRequest.getParameter("folderName")).thenReturn("Folder1");
+    when(mockRequest.getParameter("folderDefaultLanguage")).thenReturn("en");
+    when(mockRequest.getParameter("userKey")).thenReturn(userKey);
+
+    servlet.doPost(mockRequest, mockResponse);
+    assertEquals(1, datastore.prepare(new Query("Folder").setAncestor(user.getKey())).countEntities(withLimit(10)));
+  }
+
+  @Test
+  public void UserCreatesAFolderAndHasMultipleFoldersAlready() throws Exception {
+    /* Return the current number of folders that user has after creating a folder */
+
+    // Generate testing folders to store in datastore
+    Folder folderA = new Folder("FIRSTFOLDER", "en");
+    Folder folderB = new Folder("SECONDFOLDER", "en");
+    
+    // Generate testing User
+    Entity user = new Entity("User", "testId");
+    String userKey = KeyFactory.keyToString(user.getKey());
+    
+    List<Folder> foldersInDatastore = EntityTestingTool.populateDatastoreWithFolders(folderA, folderB, datastore, userKey);
+
+    when(mockRequest.getParameter("folderName")).thenReturn("Folder1");
+    when(mockRequest.getParameter("folderDefaultLanguage")).thenReturn("en");
+    when(mockRequest.getParameter("userKey")).thenReturn(userKey);
+
+    servlet.doPost(mockRequest, mockResponse);
+    assertEquals(3, datastore.prepare(new Query("Folder").setAncestor(user.getKey())).countEntities(withLimit(10)));
+
+  }
+
+  public Map<String, Object> getExpectedJsonInfo(List<Folder> folders, boolean showCreateFormStatus) {
```

Comment:
Per the testing best practices — I would strongly suggest [inlining](https://en.wikipedia.org/wiki/Inline_expansion) this function into each method.

Reply:
(because right now showCreateFormStatus and userFolders are essentially test-global constants that are not obvious or implied by the name of the `getExpectedJsonInfo` method.)

Reply:
Actually, even better would be to write out the expected JSON explicitly in each test. That way you don't rely on Gson to encode your response properly (what if Gson has a bug? Then this test won't catch it, as both this test and the servet under test use Gson to encode, so this test will continue to succeed.)

Reply:
very good point. thanks